src/app/common/UpgradeCheck.spec.ts

```diff
@@ -1,29 +1,36 @@
 import { describe, expect, test } from '@jest/globals'
 import { flushPromises, mount } from '@vue/test-utils'
+import { container, TOKENS, service, injected } from '@/services'
+import Env from '@/services/env/Env'
 
 import UpgradeCheck from './UpgradeCheck.vue'
-import container from '@/services/container'
 
 function renderComponent() {
   return mount(UpgradeCheck)
 }
 
 describe('UpgradeCheck.vue', () => {
   test('renders snapshot', async () => {
-    const env = container.get('env')
-    container.set('env', {
-      var: (key: string) => {
+    container.capture?.()
+
+    class TestEnv extends Env {
+      var(...rest: Parameters<Env['var']>) {
+        const key = rest[0]
         if (key === 'KUMA_VERSION') {
           return '1.2.0'
         }
-        return env.var(key)
-      },
-    })
+        return super.var(...rest)
+      }
+    }
+    TOKENS.Env = service(TestEnv, { description: 'Env' })
+    injected(TestEnv, TOKENS.EnvVars)
```

Comment:
**Comment**: This is very complicated now.

Previously this was achieved using:

```js
store.state.config.version = '1.2.0'
```

From what I can tell, we seem to be replacing the _entire_ `Env` utility with a custom one specific to this test. Is that correct? I had really hoped for this to be simpler.

One thing that confuses me is that the symbols `service` and `injected` are very unclear in their naming. What is `service` doing? Constructing a service? Returning an existing service? Same for `injected`. The names of those functions do not indicate what they’re doing.

I’d appreciate a slim piece of documentation in the project that explains how one makes use of ... this stuff. And what things are called (e.g. what’s a service? what’s a token? etc.).

Reply:
This is just overriding a method on a class in order to provide a mock, I personally don't think that's a very complicated thing, and probably pretty common. Is this a blocker, or just an opinion?

It's almost:

`env.var = () => '1.2.0'`

> From what I can tell, we seem to be replacing the entire Env utility with a custom one specific to this test. Is that correct?

Yes, that's correct.

`injected` comes from brandi which we've said we are going to use as a service container. Docs for `injected` specifically are here https://brandi.js.org/reference/pointers-and-registrators#injectedtarget-tokens.

`service` is a thin helper function the following (`service` is easier to write than the following, especially when the vast majority of our services will be singletons):

```javascript
const TOKENS = {
  apiService: token<ApiService>('ApiService') 
};
container.bind(TOKENS.apiService).toInstance(ApiService).inSingletonScope();
```

https://brandi.js.org/reference/binding-scopes#insingletonscope

A `token` is an identifier for a `service` that carries its type with it instead of just using strings which would mean you lose the type.

https://brandi.js.org/reference/pointers-and-registrators#tokentdescription

All the docs for all this stuff is at https://brandi.js.org/, and if we want to keep the thin helper functions, then I'd be happy to add some brief docs for those if it would help.

> And what things are called (e.g. what’s a service? what’s a token? etc.).

Take a look at brandi's docs, a lot of this was covered in the MADR along with links to these things.

Reply:
> This is just overriding a method on a class in order to provide a mock, I personally don't think that's a very complicated thing, and probably pretty common.

Absolutely speaking, it’s not very complicated, but compared to what we did before, it is very much more complicated.

> Is this a blocker, or just an opinion?

I label my PR comments very clearly (following roughly what’s outlined in https://conventionalcomments.org/) and so this is a comment.

> It's almost:
>
> env.var = () => '1.2.0'

But it’s really not and it’s not close. If it were that, it wouldn’t seem complicated to do, but as it is, the amount of boilerplate needed for mocking stuff is undoubtedly much more complicated.

---

Thanks for writing up some explanations on those concepts. I understand where these symbols are coming from.

So, in summary `service` retrieves a service and `injected` is ... registering that service again because it was changed? If so, In my opinion, `injected` is a terrible name.

> All the docs for all this stuff is at https://brandi.js.org/, and if we want to keep the thin helper functions, then I'd be happy to add some brief docs for those if it would help.

I don’t think we should be asking people who onboard in this project that they need to be learning Brandi. Not only is this something we said we definitely want to avoid by hiding away even the fact that we are using brandi (or whichever library we use for DI), but it also asks for a degree of understanding of the intricacies of Brandi that really, really shouldn’t be required for a project like Kuma GUI.

What I’m looking for is a short primer that tells me or a new developer how to create a new service and how to use an existing one and how to change one in a context where it needs to be changed.

---

As a point of personal opinion, I want to know as little as I need to be able to use these utilities. I understand at which points we need this kind of complexity, but for the rest of the day-to-day programming activities in this project, I rather keep things simple.

Reply:
Thanks again for clarifying some of the terms used here.

If you don’t think documentation for the use of the DI pattern in Kuma GUI, that’s fine by me. I think that would be pretty sensible seeing that it’s not a pattern many people will be overtly familiar with.

Reply:
No problem, I added more documentation around this in the root of the `@/services` folder just incase (see https://github.com/kumahq/kuma-gui/blob/e15711deda3b51b1aecfc2e93ff79ad37ff31764/src/services/README.md). 

Take into account that the MADR as plenty of info and links to more info also.